Ideally to run this notebook you should have a recent enough version of Python3 with numpy installed.

### Q1-Q4

The cell below is just the code with the functions, results are printed out in Q6 further down.

In [4]:
import numpy as np

K = 10
T = 10_000
L_ODD = np.array([1,-1,*[0.1]*(K-2)])
L_EVEN = np.array([-1,1,*[0.1]*(K-2)])

# Q1 and Q3:
def ftl():
    '''this fn returns the ftl regret'''
    cumul_losses = np.zeros(K)
    ftl_loss = 0

    for t in range(1, T):
        I_t = np.argmin(cumul_losses)
        if t%2==0:
            ftl_loss += L_EVEN[I_t]
            cumul_losses += L_EVEN
        else:
            ftl_loss += L_ODD[I_t]
            cumul_losses += L_ODD
    regret = ftl_loss - min(cumul_losses)
    return regret

# Q2 and Q4:
def exp_weights(seed):
    '''returns the exponential weights regret'''
    cumul_losses = np.zeros(K)
    eta = np.sqrt(2*np.log(K)/T)
    experts = np.arange(K)
    p_arr = np.array([1/K]*K)
    rng = np.random.default_rng(seed=seed)
    exp_weight_loss = 0

    for t in range(1, T):
        I_t = rng.choice(experts, size=1, p=p_arr)
        if t%2==0:
            exp_weight_loss += L_EVEN[I_t]
            cumul_losses += L_EVEN
        else:
            exp_weight_loss += L_ODD[I_t]
            cumul_losses += L_ODD

        max_j = eta*max(cumul_losses)
        common_denom = max_j - np.log(np.sum(np.exp(-eta*cumul_losses+max_j)))
        p_arr = np.exp(-eta * cumul_losses + common_denom)

    regret = exp_weight_loss - min(cumul_losses)
    return regret

### Q5
From the lecture notes we have:
$\sum_t\sum_i p_t(i)l_t(i) - \min_i \sum_t l_t(i) \leq L\sqrt{2\ln(K)T}$

The lecture notes also mention the Azuma Hoeffding inequality:

$\sum_t l_t(I_t) - \min_i \sum_t l_t(i) \leq \sum_t\sum_i p_t(i)l_t(i) - \min_i \sum_t l_t(i) + 2L\sqrt{T\ln{1/\delta}}$

with probability at least $1-\delta$. so by simply combining the two above inequalities we can write

$\sum_t l_t(I_t) - \min_i \sum_t l_t(i) \leq \sum_t\sum_i p_t(i)l_t(i) - \min_i \sum_t l_t(i) + 2L\sqrt{T\ln{1/\delta}} \leq L\sqrt{2\ln(K)T} + 2L\sqrt{T\ln{1/\delta}}$

and keeping the right and left sides:

$\sum_t l_t(I_t) - \min_i \sum_t l_t(i) \leq L\sqrt{2\ln(K)T} + 2L\sqrt{T\ln{1/\delta}}$

, where the left term is the $\text{regret}$.

Like we said this inequality hold with probability $1-\delta$, so there is always a $\delta$ probability that this inequality does not hold, which means $P\left(\text{regret}>L\sqrt{2\ln(K)T} + 2L\sqrt{T\ln{1/\delta}}\right) > 0$ unless $\delta$ is $0$, but in that case the second term on the right would be $+\infty$, which is strictly positive so the right expression in the inequality would be strictly greater than $L\sqrt{2\ln(K)T}$ anyway. So in no scenario are we guaranteed that the regret will be bounded by $L\sqrt{2\ln(K)T}$, unless we only have 1 expert ($K=1$) which would defeat the purpose of the algorithm anyway.

### Q6

In [ ]:
v_ftl = ftl()
seeds = np.arange(20)
avg_v_ew = np.mean([exp_weights(seed) for seed in seeds])
print('V_TFL = ', round(v_ftl,3))
print('average V_EW = ', round(avg_v_ew,3))

V_TFL =  10000.0
average V_EW =  145.2


Based on this 20-trial experiment I would of course pick the EW algorithm as it yields an expected regret that is much lower than that of this deterministic version of Ftl.

### Q7-Q8

In [ ]:
def sample_losses():
    x_1,x_2 = np.random.binomial(n=1, p=.5, size=2)
    losses = np.array([x_1-.9*(1-x_1),2*x_2-1,*[0.1]*(K-2)])
    return losses

def bernoulli_ftl():
    '''this fn returns the ftl regret'''
    cumul_losses = np.zeros(K)
    ftl_loss = 0

    for t in range(1, T):
        I_t = np.argmin(cumul_losses)
        losses = sample_losses()
        ftl_loss += losses[I_t]
        cumul_losses += losses
    regret = ftl_loss - min(cumul_losses)
    return [ftl_loss, regret]

def bernoulli_exp_weights():
    '''returns the exponential weights regret'''
    cumul_losses = np.zeros(K)
    eta = np.sqrt(2*np.log(K)/T)
    experts = np.arange(K)
    p_arr = np.array([1/K]*K)
    rng = np.random.default_rng()
    exp_weight_loss = 0

    for t in range(1, T):
        I_t = rng.choice(experts, size=1, p=p_arr)
        losses = sample_losses()
        exp_weight_loss += losses[I_t].item()
        cumul_losses += losses

        max_j = eta*max(cumul_losses)
        common_denom = max_j - np.log(np.sum(np.exp(-eta*cumul_losses+max_j)))
        p_arr = np.exp(-eta * cumul_losses + common_denom)

    regret = exp_weight_loss - min(cumul_losses).item()
    return [exp_weight_loss, regret]

In [ ]:
# Q7+Q8:

l_ftl, v_ftl = np.array([bernoulli_ftl() for _ in range(20)]).mean(0).round(3)
l_ew, v_ew = np.array([bernoulli_exp_weights() for _ in range(20)]).mean(0).round(3)
print('average L_FTL = ', l_ftl, ', average V_FTL = ', v_ftl)
print('average L_EW = ', l_ew, ', average V_EW = ', v_ew)

average L_FTL =  6.08 , average V_FTL =  15.98
average L_EW =  125.47 , average V_EW =  118.67


### Q9

The loss incurred with this stochastic version of FTL is substantially lower than the EW loss (even when I repeated this 20-repetition experiment many times). Since my goal is to minimise regret, I would therefore pick this stochastic version of FTL over EW here.

### Q10

We cannot guarantee that. One of the assumptions of Theorem 2.4.3 in the lecture notes is that $\max_{i,t}|l_t(i)|\leq L$, but since we're sampling the losses $l_t(3)$ and $l_t(4)$ from the normal distro, the latter's domain spans the whole real axis so those losses can technically be infinity. That theorem is therefore unusable because those two losses and thus the expected regret are not guaranteed to be bounded.

### Sticker contest

Since I could not think of a revolutionary algorithm in the span of 2-3 days I decided to play around with the learning rate in EW. I noticed that in the classic EW the learning rate stays constant. I remembered that in research-grade deep learning the learning rate is usually decreased with a schedule, so I was curious to try it here and it's straightforward to implement. The two Z losses are deterministic, and we also know that the normal distributions that we're sampling the two X losses from are stationary with mean $\mu$, stdev of 1. SO since we don't expect those distributions to shift during the experiment, there shouldn't be an unexpected change of behaviour during the experiment that would require us to crank our learning rate back up to find a new changed global minimum. At least that's how i imagine things here. So I opted to try cosine annealing (and since there's no distro shift we shouldn't need restarts).

To change the mean of the normal distributions, change the "loc" argument in `sample_norm_losses` below (scale is the stdev).

In [ ]:
K=10
T=10_000

def sample_norm_losses():
    x_3 = np.random.normal(loc=0., scale=1, size=1).item()
    x_4 = np.random.normal(loc=0., scale=1, size=1).item()
    return x_3, x_4

def sticker_exp_weights(anneal=True, eta_min_factor=0.):
    '''returns the loss of the algo'''
    cumul_losses = np.zeros(K)
    experts = np.arange(K)
    p_arr = np.array([1/K]*K)
    rng = np.random.default_rng()
    exp_weight_loss = 0

    # eta max is the lr in the classic EW algo
    eta_max = np.sqrt(2*np.log(K)/T)
    eta_min = eta_max*eta_min_factor
    eta = eta_max #useless in case we anneal but just to remove the unbound warning...

    for t in range(T):
        # i chose the cosine annealing formula used by pytorch:
        if anneal:
            eta = eta_min+(eta_max-eta_min)*(1+np.cos(t*np.pi/T))

        I_t = rng.choice(experts, size=1, p=p_arr).item()

        z_1, z_2 = (-1, 1) if t%2 == 0 else (1, -1)
        x_3, x_4 = sample_norm_losses()
        losses = [z_1, z_2, x_3, x_4, *[0.1]*(K-4)]
        exp_weight_loss += losses[I_t]
        cumul_losses += losses

        max_j = eta*max(cumul_losses)
        common_denom = max_j - np.log(np.sum(np.exp(-eta*cumul_losses+max_j)))
        p_arr = np.exp(-eta * cumul_losses + common_denom)

    # regret = exp_weight_loss - min(cumul_losses).item()
    return exp_weight_loss

While doing this assignment i noticed the average of 20 trials still has very high variance, so unless I found a mindblowingly superior algorithm, one 20-trial experiment will not do justice to the potential difference between my algo and the classic EW.

So i decided to design a quick Monte Carlo experiment as we were taught in computational statistics: I repeat the 20-trial experiment `MC_reps` times, so here 30 times for example, then take the average of the 30 average losses and its Monte Carlo stdev which is the sample stdev (ddof=1) divided by the square root of the sample size which is 30. So following the CLT we can expect the true average perf of my algo over 20 trials to fall within $\pm1.96\times$ Monte carlo stdev.

I also run the MC experiment for the classic EW (no lr annealing) and compare the result to my algo.

Note: the `eta_min_factor` of $.25$ means the minimum learning rate reached at round $T$ is $.25\times$ the initial learning rate (which I set to the learning rate of the classic EW), and this value was found to yield better results after trying a few values between $0.01$ and $0.5$.

Note 2: the cell below ran in ~2min40s on a Macbook neo laptop with an apple a18pro chip.

In [44]:
MC_reps = 30

avg_ew = []
avg_stud = []
for r in range(MC_reps):
    l_ew = [sticker_exp_weights(anneal=False) for _ in range(20)]
    l_stud = [sticker_exp_weights(anneal=True, eta_min_factor=0.25) for _ in range(20)]
    avg_ew.append(np.mean(l_ew))
    avg_stud.append(np.mean(l_stud))

# avg of mc estimator:
avg_l_ew = np.mean(avg_ew).round(3)
# mc stdev:
sd_l_ew = round(np.std(avg_ew, ddof=1)/np.sqrt(MC_reps),3)
avg_l_stud = np.mean(avg_stud).round(3)
sd_l_stud = round(np.std(avg_stud, ddof=1)/np.sqrt(MC_reps),3)
print('average L_EW = ', avg_l_ew, '+-', sd_l_ew)
print('average L_stud = ', avg_l_stud, '+-', sd_l_stud)

average L_EW =  89.234 +- 5.339
average L_stud =  72.791 +- 4.436


My algo scored a $72.791$ average loss against $89.234$ for classic EW. Naturally one will get slightly different results when the cell is rerun because of randomness.

Not bad. Perhaps this result could be improved by running a finer grid search on the minimum learning rate or trying a different annealing algo.